plumbing/protocol/packp/sideband/demux.go

```diff
@@ -15,21 +14,22 @@ var ErrMaxPackedExceeded = errors.New("max. packed size exceeded")
 // Progress allows to read the progress information
 type Progress interface {
 	io.Reader
+	io.Writer
 }
 
 // Demuxer demultiplex the progress reports and error info interleaved with the
 // packfile itself.
 //
-// A sideband has three diferent channels the main one call PackData contains
+// A sideband has three diferent channels the main one, call PackData, contains
 // the packfile data, the ErrorMessage channel, that contains server errors and
 // the last one ProgressMessage channel containing information about the ongoing
 // tast happening in the server (optinal, can be suppressed sending NoProgress
 // or Quiet capabilities to the server)
 //
 // In order to demultiplex the data stream, method `Read` should be called to
 // retrieve the PackData channel, the incoming data from the ProgressMessage is
-// stored and can be read from `Progress` field, if any message is retrieved
-// from the ErrorMessage channel an error is returned and we can assume that the
+// written at `Progress` (if any), if any message is retrieved from the
+// ErrorMessage channel an error is returned and we can assume that the
 // conection has been closed.
 type Demuxer struct {
```

Comment:
This is confusing as the Demuxer type does not Demultiplex the channels and it is not clear how are errors messages sent back to the user.  Also a connection is mentioned in the comment, and some assumptions about it being closed, but it is not clear how this can be relevant for anything.


I think a better comment would be something like this:

```
[Foo] is a reader to the packfile data in the PackData channel.  The progress information
is writen to the the Progress field while reading.  If any message is received through the
ErrorMessage channel, the `Read` method will return it as an error and it will assume the
connection has been closed.
```

Although I am still not sure about the connection thing.

I would rather use a real demuxer though or even simply a method that returns 3 readers (or even better, 3 scanners): one for each channel, treating them as equal, with the same usage semantics.  This will have a much simpler API and will let the user decide how to use each channel.

Another approach would be to use a single scanner, returning channel and data when read.